File: 搜索与回溯/494.py
from typing import List
from functools import lru_cache


class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        @lru_cache(None)
        def dfs(idx,t):
            if idx == len(nums):
                return 1 if t==0 else 0
            return dfs(idx+1,t-nums[idx]) + dfs(idx+1,t+nums[idx])
        return dfs(0,target)

nums = [1, 1, 1, 1, 1]
target = 3
obj = Solution()
print(obj.findTargetSumWays(nums, target))

File: 搜索与回溯/494.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int count = 0;
    int findTargetSumWays(vector<int> &nums, int target)
    {
        dfs(nums, 0, 0, target);
        return count;
    }

    void dfs(vector<int> &nums, int idx, int Sum, int target)
    {
        if (idx == nums.size())
        {
            if (Sum == target)
            {
                count++;
            }
        }
        else
        {
            dfs(nums, idx + 1, Sum + nums[idx], target);
            dfs(nums, idx + 1, Sum - nums[idx], target);
        }
    }
};